astropy/table/table.py

```diff
@@ -3270,7 +3273,12 @@ def from_pandas(cls, dataframe, index=False):
             datas.insert(0, np.array(dataframe.index))
             masks.insert(0, np.zeros(len(dataframe), dtype=bool))
 
-        for name, column, data, mask in zip(names, columns, datas, masks):
+        if units is None:
```

Comment:
Before this point you should validate that all the provided column names in `units` are in the table.
```
for name in units:
    if name not in names:
       raise KeyError(f'units column name {name} is not valid')
```
Then add a test.

Reply:
Do we really want to raise an error if the dict has additionial names in it?

Reply:
Maybe just a warning so typos are easy to spot?

Reply:
Sure, a warning is OK.  I think the most likely reason for a name in the `units` dict not being in the table is a program error / typo, so it is important not just silently ignore.

Reply:
Added a Check for Mapping and warnings if additional columns are encountered